### PR TITLE
dialects: Add filecheck tests for stencil dialect and init rewriting

### PR DESCRIPTION
**Pull request by meshtag**
This PR intends to introduce some filecheck tests for the stencil dialect and init rewriting infra. My motive behind this is to add a reference for related stencil IR and smooth things out for rewriting. I have taken the liberty to change the definition of some ops to make them workable. This should also serve as a starting point for peep-hole rewrites of stencil ops. 

Operation semantics are inspired from [here](https://github.com/spcl/open-earth-compiler/blob/master/include/Dialect/Stencil/StencilOps.td). Filecheck tests are inspired from [here](https://github.com/spcl/open-earth-compiler/blob/master/test/Examples/copy.mlir) and [here](https://github.com/spcl/open-earth-compiler/blob/master/test/Examples/laplace.mlir).

Additionally, I have also introduced a small rewrite for changing `stencil.field` types in `FuncOp`. 

This PR depends on #497 .

### CODE REVIEW

File: xdsl/dialects/experimental/stencil.py [hunk 1]
```diff
@@ -336,16 +339,16 @@ class Combine(Operation):
     lower_ext: Annotated[VarOperand, TempType]
     upper_ext: Annotated[VarOperand, TempType]
 
-    lb = OptOpAttr[Stencil_Index]
-    ub = OptOpAttr[Stencil_Index]
+    # lb = OptOpAttr[Stencil_Index]
```
Comment (superlopuh) [inline xdsl/dialects/experimental/stencil.py:342]: nit: commented code

**Comment by webmiche** [comment]
> I am not sure if `CI-MLIR-based Testing` is failing due to changes made in this PR. The error we get is this
> 
> ```
> Error: fatal: unable to access 'https://github.com/xdslproject/xdsl/': The requested URL returned error: 429
> Error: The process '/usr/bin/git' failed with exit code 128
> ```

In these cases, you can just manually retrigger the failed jobs. Sometimes that fixes the issue.

**Comment by superlopuh** [comment]
It looks like there are some more pyright issues remaining, let me know if you'd like help setting up your environment to hunt them down :) The command to test what the CI will spit out is `pyright -p pyright-ci.json`

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    return MemRefType.from_params(inputFieldType.element_type,
+                                  memref_shape_array_attr)
+
+
+def GetMemRefFromFieldWithLBAndUB(memref_element_type: _TypeElement,
+                                  lb: IndexType,
+                                  ub: IndexType) -> MemRefType[_TypeElement]:
+    # Assumes lb and ub are of same size and same underlying element types.
+    memref_shape_integer_attr_list = []
```
Comment (math-fehr) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:35]: You may want to add a type here to make pyright pass.

**Review by math-fehr — APPROVED** [review]
Nice!
I added just some comments, but overall since this is experimental I did not review it too much in details!

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    for i in range(len(lb.parameters[0].data)):
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:36]: I think it comes from some confusion in the MLIR dialect.
It is defined as `def Stencil_Index : Confined<I64ArrayAttr, [ArrayCount<3>]>;` in TableGen, I'll do something similar in here

[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py]: @math-fehr (Not urgent) I made Pyright happy, but I'm super confused by how to express `Confined<I64ArrayAttr, [ArrayCount<3>]>` in IRDL, here in an OpAttr. I went down the route of this PR, defining a custom parameterized attribute, but couldn't get to use a custom `AttrConstraint`there either, so just defined a custom `verify` at the Attribute level.
This is good-enough for now, I don't think we expect tight interoperability with OpenEarth and stencil is not upstream MLIR, but i think this raises good design questions for xDSL's IRDL API.
Do you know how to do that? (Just expressing that an op expects as attribute an array of T of lenght N. Even better if we can factorize it behind a Python object.)

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            IntegerAttr.from_params(i.value.data, i.typ))
+
+    memref_shape_array_attr = ArrayAttr[AnyIntegerAttr](
+        memref_shape_integer_attr_list)
+
+    return MemRefType.from_params(inputFieldType.element_type,
+                                  memref_shape_array_attr)
+
+
+def GetMemRefFromFieldWithLBAndUB(memref_element_type: _TypeElement,
```
Comment (webmiche) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:30]: Feels like these might be useful as accessors in memeref?

We cna also leave them here for now, they don't need tests here I guess...

**Review by webmiche — APPROVED** [review]
LGTM

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 4 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:30]: No, those IndexAttr are stencil.IndexAttr. I guess this is a point for always prefixing those with the dialect for readability though!

**Review by superlopuh — APPROVED** [review]

**Pull request merged**